pkg/controllers/disruption/consolidation.go

```diff
@@ -101,61 +97,129 @@ func (c *consolidation) ShouldDisrupt(_ context.Context, cn *Candidate) bool {
 		c.recorder.Publish(disruptionevents.Unconsolidatable(cn.Node, cn.NodeClaim, fmt.Sprintf("NodePool %q has consolidation disabled", cn.nodePool.Name))...)
 		return false
 	}
-	return true
-}
-
-// sortCandidates sorts candidates by disruption cost (where the lowest disruption cost is first) and returns the result
-func (c *consolidation) sortCandidates(candidates []*Candidate) []*Candidate {
-	sort.Slice(candidates, func(i int, j int) bool {
-		return candidates[i].disruptionCost < candidates[j].disruptionCost
-	})
-	return candidates
+	return cn.NodeClaim.StatusConditions().GetCondition(v1beta1.Underutilized).IsTrue()
 }
 
-// computeConsolidation computes a consolidation action to take
+// ComputeCommand computes a consolidation action to take
 //
 // nolint:gocyclo
-func (c *consolidation) computeConsolidation(ctx context.Context, candidates ...*Candidate) (Command, pscheduling.Results, error) {
-	var err error
-	// Run scheduling simulation to compute consolidation option
-	results, err := SimulateScheduling(ctx, c.kubeClient, c.cluster, c.provisioner, candidates...)
-	if err != nil {
-		// if a candidate node is now deleting, just retry
-		if errors.Is(err, errCandidateDeleting) {
-			return Command{}, pscheduling.Results{}, nil
+func (c *Consolidation) ComputeCommand(ctx context.Context, disruptionBudgetMapping map[string]int, candidates ...*Candidate) (Command, pscheduling.Results, error) {
+	// All the candidates here all have the v1beta1.Underutilized condition
+	commands := c.getActiveValidConsolidations(ctx, candidates...)
+	for _, cmd := range commands {
+		// Run scheduling simulation to compute consolidation option
+		results, err := SimulateScheduling(ctx, c.kubeClient, c.cluster, c.provisioner, cmd.candidates...)
+		if err != nil {
+			// if a candidate node is now deleting, just retry
+			if errors.Is(err, errCandidateDeleting) {
+				return Command{}, pscheduling.Results{}, nil
+			}
+			return Command{}, pscheduling.Results{}, err
+		}
+		results, valid, err := validateResults(ctx, results, c.recorder, cmd.candidates...)
+		if err != nil {
+			return Command{}, pscheduling.Results{}, fmt.Errorf("validating scheduling simulation results, %w", err)
+		}
+		// If valid, we exit early and return the command
+		if valid {
+			return Command{
+				candidates:   cmd.candidates,
+				replacements: results.NewNodeClaims,
+			}, results, nil
 		}
-		return Command{}, pscheduling.Results{}, err
 	}
 
+	return Command{}, pscheduling.Results{}, nil
+}
+
+// returns a map of command-id to list of candidate nodeclaim names
+// returned sorted by creation timestamp
+func (c *Consolidation) getActiveValidConsolidations(ctx context.Context, candidates ...*Candidate) []Command {
```

Comment:
Is there a scenario where cluster state is out of sync causing us to recreate partial commands? Maybe this is fine, since if we can consolidate a set of nodes we can also consolidate a subset, and the remaining nodes would get picked up by subsequent runs of the disruption controller. It definitely feels a little icky though if that is possible, especially since we're making command UIDs observable. It may also result in additional node churn since we may create multiple smaller replacement nodeclaims when executing multiple partial commands vs a single larger nodeclaim.

Reply:
Seems like there might be some scaffolding for this in cluster state that's unused, were you able to determine it was unnecessary?